File: quattro2/win10/JavaScriptInterface/NativeCall+midi.cpp
//
//  NativeCall+midi.cpp
//
//  Copyright 2015 Roland Corporation. All rights reserved.
//

#include "NativeCall.h"

static LPCWSTR OBJ_NAME = L"midi";

enum {
	dispid_midi_in_endpoints,
	dispid_midi_in_connect,
	dispid_midi_in_disconnect,
	dispid_midi_out_endpoints,
	dispid_midi_out_connect,
	dispid_midi_out_disconnect,
	dispid_midi_send,
	dispid_midi_panel,
};

NativeCall_midi::NativeCall_midi(NativeObjects* objs) : NativeCall(objs)
{
	set(L"inendpoints",   dispid_midi_in_endpoints);
	set(L"inconnect",     dispid_midi_in_connect);
	set(L"indisconnect",  dispid_midi_in_disconnect);
	set(L"outendpoints",  dispid_midi_out_endpoints);
	set(L"outconnect",    dispid_midi_out_connect);
	set(L"outdisconnect", dispid_midi_out_disconnect);
	set(L"send",          dispid_midi_send);
	set(L"panel",         dispid_midi_panel);
}

LPCWSTR NativeCall_midi::InterfaceName() const { return OBJ_NAME; }

void NativeCall_midi::Dispatch(UINT dispid, const CefV8ValueList& args, CefRetValue& ret)
{
	MIDIClient* midi = mp_objs->midi;

	switch (dispid) {

		case dispid_midi_in_endpoints:
		{
			ret = JSON::stringify(midi->inputPort()->endpoints());
			break;
		}
		case dispid_midi_in_connect:
		{
			if (args.size() > 0) {
				MIDI_ENDPOINT_INFO ep = JSON::endpoint(args[0]->GetStringValue().c_str());
				midi->inputPort()->connectEndpoint(&ep);
			} else {
				midi->inputPort()->connectAllEndpoints();
			}
			break;
		}
		case dispid_midi_in_disconnect:
		{
			if (args.size() > 0) {
				MIDI_ENDPOINT_INFO ep = JSON::endpoint(args[0]->GetStringValue().c_str());
				midi->inputPort()->disconnectEndpoint(&ep);
			} else {
				midi->inputPort()->disconnectAllEndpoints();
			}
			break;
		}
		case dispid_midi_out_endpoints:
		{
			ret = JSON::stringify(midi->outputPort()->endpoints());
			break;
		}
		case dispid_midi_out_connect:
		{
			if (args.size() > 0) {
				MIDI_ENDPOINT_INFO ep = JSON::endpoint(args[0]->GetStringValue().c_str());
				midi->outputPort()->connectEndpoint(&ep);
			} else {
				midi->outputPort()->connectAllEndpoints();
			}
			break;
		}
		case dispid_midi_out_disconnect:
		{
			if (args.size() > 0) {
				MIDI_ENDPOINT_INFO ep = JSON::endpoint(args[0]->GetStringValue().c_str());
				midi->outputPort()->disconnectEndpoint(&ep);
			} else {
				midi->outputPort()->disconnectAllEndpoints();
			}
			break;
		}
		case dispid_midi_send:
		{
			std::string data = HEX::data(args[0]->GetStringValue().c_str());
			midi->outputPort()->send((BYTE*)data.c_str(), (int)data.length());
			break;
		}
		case dispid_midi_panel:
		{
			::ShellExecute(NULL, NULL, TEXT("ms-settings:bluetooth"), NULL, NULL, SW_SHOWNORMAL);
			break;
		}

	}
}

void _MIDIClientDelegate::midiInputPortConnectFailed(const MIDI_ENDPOINT_INFO* endpoint)
{
	CStringW ev;
	ev.Format(L"%s\f%s\f%s", OBJ_NAME, L"connectfailed", JSON::stringify(endpoint));
	mp_intf->postEvent(ev);
}

void _MIDIClientDelegate::midiOutputPortConnectFailed(const MIDI_ENDPOINT_INFO* endpoint)
{
	CStringW ev;
	ev.Format(L"%s\f%s\f%s", OBJ_NAME, L"connectfailed", JSON::stringify(endpoint));
	mp_intf->postEvent(ev);
}

void _MIDIClientDelegate::midiObjectAddedRemoved()
{
	CStringW ev;
	ev.Format(L"%s\f%s", OBJ_NAME, L"changed");
	mp_intf->postEvent(ev);
}

void _MIDIClientDelegate::midiErrorDidOccur(MMRESULT error)
{
	CStringW ev;
	ev.Format(L"%s\f%s\f%ld", OBJ_NAME, L"error", error);
	mp_intf->postEvent(ev);
}

void _MIDIClientDelegate::midiInputMessage(const BYTE* msg, int bytes, DWORD msec)
{
	CStringW ev;
	ev.Format(L"%s\f%s\f%s\f%u", OBJ_NAME, L"message", HEX::text(msg, bytes), msec);
	mp_intf->postEvent(ev);
}

File: quattro2/mac/Common/JavaScriptInterface/NativeCall+midi.m
//
//  NativeCall+midi.m
//
//  Copyright 2015 Roland Corporation. All rights reserved.
//

#import "NativeCall.h"

#import <TargetConditionals.h>
#if !TARGET_OS_IPHONE
#import <AppKit/AppKit.h>
#endif

#ifdef USE_BLEMIDI_SCANNER
  #if TARGET_OS_IPHONE && defined(USE_VENDER_BLEMIDI)
    #import "../../iOS/VenderBLEMIDI/BLEMIDIConnectionManager.h"
    @interface DummyTableView : UITableView
    - (void)reloadData;
    @end
    static DummyTableView *_tableView = nil;
  #else
    #undef USE_BLEMIDI_SCANNER
  #endif
#endif

static NSString *const OBJ_NAME = @"midi";

@interface NativeCall (midi) @end

@implementation NativeCall (midi)

- (NSString *)$$midi_inendpoints:(NSString *)param1 :(NSString *)param2
{
	return_s(JSONStringWithObject([midi.inputPort endpoints]));
}

- (NSString *)$$midi_inconnect:(NSString *)param1 :(NSString *)param2
{
	if (param1) {
		[midi.inputPort connectEndpoint:dictionaryWithJSONString(param1)];
	} else {
		[midi.inputPort connectAllEndpoints];
	}
	undefined;
}

- (NSString *)$$midi_indisconnect:(NSString *)param1 :(NSString *)param2
{
	if (param1) {
		[midi.inputPort disconnectEndpoint:dictionaryWithJSONString(param1)];
	} else {
		[midi.inputPort disconnectAllEndpoints];
	}
	undefined;
}

- (NSString *)$$midi_outendpoints:(NSString *)param1 :(NSString *)param2
{
	return_s(JSONStringWithObject([midi.outputPort endpoints]));
}

- (NSString *)$$midi_outconnect:(NSString *)param1 :(NSString *)param2
{
	if (param1) {
		[midi.outputPort connectEndpoint:dictionaryWithJSONString(param1)];
	} else {
		[midi.outputPort connectAllEndpoints];
	}
	undefined;
}

- (NSString *)$$midi_outdisconnect:(NSString *)param1 :(NSString *)param2
{
	if (param1) {
		[midi.outputPort disconnectEndpoint:dictionaryWithJSONString(param1)];
	} else {
		[midi.outputPort disconnectAllEndpoints];
	}
	undefined;
}

- (NSString *)$$midi_send:(NSString *)param1 :(NSString *)param2
{
    NSData *data = dataWithHexString(param1);
    [midi.outputPort send:(u_int8_t *)data.bytes size:(int)data.length];
	undefined;
}

- (NSString *)$$midi_panel:(NSString *)param1 :(NSString *)param2
{
#if TARGET_OS_IPHONE
    [[NSNotificationCenter defaultCenter] postNotificationName:@"coreMIDIBLE" object:nil userInfo:nil];
#else
    NSURL *url = [[NSWorkspace sharedWorkspace] URLForApplicationWithBundleIdentifier:@"com.apple.audio.AudioMIDISetup"];
    [[NSWorkspace sharedWorkspace] openURL:url];
#endif
	undefined;
}

#ifdef USE_BLEMIDI_SCANNER
- (NSString *)$$blemidi_scanstart:(NSString *)param1 :(NSString *)param2
{
	if (!_tableView) {
        _tableView = [[DummyTableView alloc] init];
    }
    [[BLEMIDIConnectionManager sharedInstance] startScan:_tableView];
    undefined;
}
- (NSString *)$$blemidi_scanstop:(NSString *)param1 :(NSString *)param2
{
    [[BLEMIDIConnectionManager sharedInstance] stopScan];
	undefined;
}
- (NSString *)$$blemidi_connect:(NSString *)param1 :(NSString *)param2
{
    NSArray *devices = [[BLEMIDIConnectionManager sharedInstance] devices];
    for (NSUInteger index = 0; index < devices.count; index++) {
        CBPeripheral *peripheral = [devices objectAtIndex:index];
        if ([param1 isEqualToString:[peripheral.identifier UUIDString]]) {
            if (peripheral.state == CBPeripheralStateDisconnected) {
                [[BLEMIDIConnectionManager sharedInstance] tapAtIndex:index];
            }
            break;
        }
    }
    undefined;
}
- (NSString *)$$blemidi_disconnect:(NSString *)param1 :(NSString *)param2
{
    NSArray *devices = [[BLEMIDIConnectionManager sharedInstance] devices];
    for (NSUInteger index = 0; index < devices.count; index++) {
        CBPeripheral *peripheral = [devices objectAtIndex:index];
        if ([param1 isEqualToString:[peripheral.identifier UUIDString]]) {
            if (peripheral.state != CBPeripheralStateDisconnected) {
                [[BLEMIDIConnectionManager sharedInstance] tapAtIndex:index];
            }
            break;
        }
    }
    undefined;
}
#endif

@end

@implementation _MIDIClientDelegate

- (void)midiInputPortConnectFailed:(NSDictionary*)endpoint
{
	NSString *event = [NSString stringWithFormat:@"%@\f%@\f%@", OBJ_NAME, @"connectfailed", JSONStringWithObject(endpoint)];
    [self.native postEvent:event];
}

- (void)midiOutputPortConnectFailed:(NSDictionary*)endpoint
{
	NSString *event = [NSString stringWithFormat:@"%@\f%@\f%@", OBJ_NAME, @"connectfailed", JSONStringWithObject(endpoint)];
    [self.native postEvent:event];
}

- (void)midiObjectAddedRemoved
{
	NSString *event = [NSString stringWithFormat:@"%@\f%@", OBJ_NAME, @"changed"];
    [self.native postEvent:event];
}

- (void)midiErrorDidOccur:(NSError*)error
{
	NSString *event = [NSString stringWithFormat:@"%@\f%@\f%ld", OBJ_NAME, @"error", (long)error.code];
    [self.native postEvent:event];
}

- (void)midiInputMessage:(const u_int8_t *)msg size:(int)bytes timeStamp:(u_int32_t)msec;
{
	NSString *hexString = hexStringWithBytes(msg, bytes);
	NSString *event = [[NSString alloc] initWithFormat:@"%@\f%@\f%@\f%u", OBJ_NAME, @"message", hexString, msec];

    [self.native postEvent:event];
	
	[hexString release];
	[event release];
}

@end

#ifdef USE_BLEMIDI_SCANNER
@implementation DummyTableView
- (void)reloadData
{
    NSMutableArray *array = [[[NSMutableArray alloc] init] autorelease];

    NSArray *devices = [[BLEMIDIConnectionManager sharedInstance] devices];
    for (CBPeripheral *peripheral in devices) {
        NSDictionary *dict = [NSDictionary dictionaryWithObjectsAndKeys:
            [peripheral.identifier UUIDString], @"id",
            peripheral.advertisingName, @"name",
            [NSNumber numberWithInteger:peripheral.state], @"state",
            nil];
        [array addObject:dict];
    }

    NSString *event = [NSString stringWithFormat:@"%@\f%@\f%@", OBJ_NAME, @"ble", JSONStringWithObject(array)];
    [[NSNotificationCenter defaultCenter] postNotificationName:POST_EVENT object:event userInfo:nil];
}
@end
#endif
